File: 0031-next-permutation/0031-next-permutation.py
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """

        pivot = -1

        for i in range(len(nums)-1, 0,-1):
            if nums[i-1]<nums[i]:
                pivot = i
                break

        if pivot == -1:
            nums.sort()
            return
        
        # print(pivot)
        for i in range(len(nums)-1,pivot-1,-1):
            if nums[pivot-1]<nums[i]:
                # print(i)
                nums[pivot-1],nums[i] = nums[i], nums[pivot-1]
                break
        nums[pivot:] = sorted(nums[pivot:])

File: 0031-next-permutation/0031-next-permutation.java
class Solution {
   public void nextPermutation(int[] nums) {
        int pivot = -1;

        // Find the pivot
        for (int i = nums.length - 1; i > 0; i--) {
            if (nums[i] > nums[i-1]) {
                pivot = i - 1;
                break;
            }
        }
        
        if (pivot == -1) {
            reverse(nums, 0, nums.length - 1);
            return;
        }

        // Find the first element that is larger than pivot
        int firstLarger = -1;
        for (int i = nums.length - 1; i > pivot; i--) {
            if (nums[i] > nums[pivot]) {
                firstLarger = i;
                break;
            }
        }

        // Swap pivot and firstLarger
        swap(nums, pivot, firstLarger);

        // Reverse the subarray after pivot
        reverse(nums, pivot + 1, nums.length - 1);
        return;
    }

    private void reverse(int[] nums, int i, int j) {
        while (i < j) {
            swap(nums, i++, j--);
        }
        return;
    }

    private void swap(int[] nums, int i, int j) {
        int tmp = nums[i];
        nums[i++] = nums[j];
        nums[j--] = tmp;
        return;
    }
}